game-core/src/main/java/games/strategy/engine/data/util/BreadthFirstSearch.java

```diff
@@ -0,0 +1,91 @@
+package games.strategy.engine.data.util;
+
+import games.strategy.engine.data.GameMap;
+import games.strategy.engine.data.Territory;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import lombok.Getter;
+import org.triplea.java.ObjectUtils;
+
+/**
+ * Implements Breadth First Search (BFS) to traverse / find territories. Since the search criteria
+ * varies depending on the use case, the class is designed to be sub-classed, with methods visit()
+ * and shouldContinueSearch() that can be overridden to customize the behavior.
+ */
+public class BreadthFirstSearch {
```

Comment:
Just a random thought I had.
This class would be super usefule to optimize `GameMap#getNeighbors(Set, Set, int, Predicate)`, which basically does the same thing here, just with less than optimal recursion.

Reply:
Yes, it's on my radar. I actually wrote the code for that, but it didn't help much with this optimization since the underlying algorithm was still bad (it was calling getNeighbors(, distance=6) on each territory). So I abandoned that to focus on the actual speed improvement which is an algorithm change.

However, when we have the new class in place, we can use it in more places like there. :)